mijia/src/bluetooth.rs

```diff
@@ -229,7 +312,153 @@ impl BluetoothSession {
+                let primary = service.primary().compat().await?;
+                services.push(ServiceInfo {
+                    id: service_id,
+                    uuid,
+                    primary,
+                });
+            }
+        }
+        Ok(services)
+    }
+
+    /// Get a list of all characteristics on the given GATT service.
+    pub async fn get_characteristics(
+        &self,
+        service: &ServiceId,
+    ) -> Result<Vec<CharacteristicInfo>, BluetoothError> {
+        let introspection_xml = self.service(service).introspect().compat().await?;
+        let service_node: Node = serde_xml_rs::from_str(&introspection_xml)?;
+        let mut characteristics = vec![];
+        for subnode in service_node.nodes {
+            let subnode_name = subnode.name.as_ref().unwrap();
+            if subnode_name.starts_with("char") {
+                let characteristic_id = CharacteristicId {
+                    object_path: format!("{}/{}", service.object_path, subnode_name),
+                };
+                let uuid = self
+                    .characteristic(&characteristic_id)
+                    .uuid()
+                    .compat()
+                    .await?;
+                characteristics.push(CharacteristicInfo {
+                    id: characteristic_id,
+                    uuid,
+                });
+            }
+        }
+        Ok(characteristics)
+    }
+
+    /// Find a GATT service with the given UUID advertised by the given device, if any.
+    ///
+    /// Note that this generally won't work until the device is connected.
+    pub async fn get_service_by_uuid(
+        &self,
+        device: &DeviceId,
+        uuid: &str,
+    ) -> Result<ServiceInfo, BluetoothError> {
+        let services = self.get_services(device).await?;
+        services
+            .into_iter()
+            .find(|service_info| service_info.uuid == uuid)
+            .ok_or_else(|| BluetoothError::uuid_not_found(uuid))
+    }
+
+    /// Find a characteristic with the given UUID as part of the given GATT service advertised by a
+    /// device, if there is any.
+    pub async fn get_characteristic_by_uuid(
+        &self,
+        service: &ServiceId,
+        uuid: &str,
+    ) -> Result<CharacteristicInfo, BluetoothError> {
+        let characteristics = self.get_characteristics(service).await?;
+        characteristics
+            .into_iter()
+            .find(|characteristic_info| characteristic_info.uuid == uuid)
+            .ok_or_else(|| BluetoothError::uuid_not_found(uuid))
+    }
+
+    /// Convenience method to get a GATT charactacteristic with the given UUID advertised by a
+    /// device as part of the given service.
+    ///
+    /// This is equivalent to calling `get_service_by_uuid` and then `get_characteristic_by_uuid`.
+    pub async fn get_service_characteristic_by_uuid(
+        &self,
+        device: &DeviceId,
+        service_uuid: &str,
+        characteristic_uuid: &str,
```

Comment:
*_uuid: &str makes me wary. We had a bunch of `uuid: uuid::Uuid` in the futurenhs project, and it caused us problems with positional arguments in the wrong order. A newtype ServiceUuid(Uuid) would avoid this class of problem.